acpi/src/madt.rs

```diff
@@ -48,7 +50,55 @@ unsafe impl AcpiTable for Madt {
     }
 }
 
+pub type PhysToVirtFn = fn(u64) -> u64;
+
 impl Madt {
+    pub fn wakeup_aps(
+        &self,
+        apic_id: u32,
+        wakeup_vector: u64,
+        phys_to_virt: PhysToVirtFn,
+    ) -> Result<(), AcpiError> {
+        let mailbox_addr = self.get_mpwk_mailbox_addr()?;
+        unsafe {
+            let mailbox = &mut *((phys_to_virt(mailbox_addr)) as *mut MultiprocessorWakeupMailbox);
+
+            mailbox.command = MpProtectedModeWakeupCommand::Noop as u16;
+
+            // Fill the mailbox
+            mailbox.apic_id = apic_id;
+            mailbox.wakeup_vector = wakeup_vector;
+            mailbox.command = MpProtectedModeWakeupCommand::Wakeup as u16;
+
+            // Wait to join
+            let timeout_loops = timeout_to_loops(Duration::from_secs(1));
+            let mut loops = 0;
+            let mut command = MpProtectedModeWakeupCommand::Wakeup;
+            while command != MpProtectedModeWakeupCommand::Noop {
```

Comment:
Rust is not guaranteed to read the actual memory address for each loop iteration without this being a volatile read. This could either use a volatile read directly, or wrappers (see [this](https://github.com/IsaacWoods/poplar/blob/main/lib/volatile/src/lib.rs) type in my OS for what I mean by this) to enforce volatile accesses.